node/src/main/scala/org/bitcoins/node/PeerManager.scala

```diff
@@ -154,7 +162,7 @@ case class PeerManager(
     //leading to a memory leak may happen
     _waitingForDeletion.add(peer)
     //now send request to stop actor which will be completed some time in future
-    client.close()
+    client.foreach(_.close())
```

Comment:
You are dropping a `Future` on the floor here. 

This introduces race conditions as the `removePeer()` method returns `Unit`, however the computation inside of `removePeer` is done asynchronously. 

Part of the bugs you are experiencing in the code you've introduced ins the fact that you aren't being consistent with handling async code. A nightmare of async code is the combination of 

1. mutability (specifically globally accessible mutability, which you are attempting to introduce here: https://github.com/bitcoin-s/bitcoin-s/pull/4456/files#diff-61413f0a707d1e3e49431ad34fcb1af04b4fa36148e75c7abc692bebc07822b7R36)
2. APIs that don't reflect asynchronous computation that occurs within a method.

This is an example of `2`. Your method return type is `Unit`, when in reality the method itself is async (`client`). 

While it may seem "easier" in the short term to use these tricks, (not reflecting async APIs, using mutability liberally), it ends in tears on machines that have multi-core processors.

Thankfully we aren't working in C which doesn't have built in abstractions (`Future`) to handle this. We are using Scala, which has the capability to express this behavior.

Lets use the tools available to us.